### PR TITLE
add web hook

### PR DESCRIPTION
**Pull request by VallarDark**

### CODE REVIEW

**Review by gitrepository100 — DISMISSED** [review]
Approved

File: OneSchedule.Domain/SetWebHookService.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+using OneSchedule.Settings;
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Telegram.Bot;
+
+namespace OneSchedule.Domain
+{
+    public class SetWebHookService : IHostedService
```
Comment (elkavitaly) [inline OneSchedule.Domain/SetWebHookService.cs]: I am not sure that this class should be placed in Domain project. It is mostly related to presentation layer, how app will receive messages. So, I think it's better to move it to OneSchedule project

File: OneSchedule.Domain/SetWebHookService.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            var assembly = AppDomain.CurrentDomain.GetAssemblies()
```
Comment (elkavitaly) [inline OneSchedule.Domain/SetWebHookService.cs]: What is the purpose of this code?

File: OneSchedule.Domain/SetWebHookService.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                .Single(o => o.EntryPoint != null);
+
+            var configuration = new ConfigurationBuilder()
```
Comment (elkavitaly) [inline OneSchedule.Domain/SetWebHookService.cs]: Isn't there any way to pass configuration using DI?

File: OneSchedule.Domain/SetWebHookService.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                .AddJsonFile("appsettings.json")
+                .AddUserSecrets(assembly)
+                .Build();
+
+            var telegramSettings = new TelegramSettings()
+            {
+                ApiKey = configuration.GetSection("TelegramSettings").GetSection("ApiKey").Value
+            };
+
+            var webHookSettings = new WebHookSettings()
```
Comment (elkavitaly) [inline OneSchedule.Domain/SetWebHookService.cs]: Try to use IOptions patters and inject these values into the constructor

File: OneSchedule.Domain/SetWebHookService.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                IsUsed = configuration.GetSection("WebHook").GetSection("IsUsed").Get<bool>(),
+                Uri = configuration.GetSection("WebHook").GetSection("Uri").Value
+            };
+
+            if (!string.IsNullOrEmpty(webHookSettings.Uri)
+                && !string.IsNullOrEmpty(telegramSettings.ApiKey))
+            {
+                var bot = new TelegramBotClient(telegramSettings.ApiKey);
+
+                if (webHookSettings.IsUsed)
```
Comment (elkavitaly) [inline OneSchedule.Domain/SetWebHookService.cs]: Use ternary operator here

File: OneSchedule.Settings/WebHookSettings.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿namespace OneSchedule.Settings
+{
+    public class WebHookSettings
+    {
+        public string Uri { get; set; }
+
+        public bool IsUsed { get; set; }
```
Comment (elkavitaly) [inline OneSchedule.Settings/WebHookSettings.cs]: I prefer to name it IsEnabled

File: OneSchedule.Domain/SetWebHookService.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            var webHookSettings = new WebHookSettings()
+            {
+                IsUsed = configuration.GetSection("WebHook").GetSection("IsUsed").Get<bool>(),
+                Uri = configuration.GetSection("WebHook").GetSection("Uri").Value
+            };
+
+            if (!string.IsNullOrEmpty(webHookSettings.Uri)
+                && !string.IsNullOrEmpty(telegramSettings.ApiKey))
+            {
+                var bot = new TelegramBotClient(telegramSettings.ApiKey);
```
Comment (elkavitaly) [inline OneSchedule.Domain/SetWebHookService.cs]: Also use DI here

**Review by elkavitaly — CHANGES_REQUESTED** [review]

File: OneSchedule/SetWebHookService.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+        public SetWebHookService(IOptions<WebHookSettings> webHookOptions, ITelegramBotClient bot )
+        {
+            _webHookOptions = webHookOptions;
+            _bot = bot;
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (!string.IsNullOrEmpty(_webHookOptions.Value.Uri) && _bot!=null)
```
Comment (elkavitaly) [inline OneSchedule/SetWebHookService.cs]: Format code

File: OneSchedule/SetWebHookService.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            {
+                if (_webHookOptions.Value.IsEnable)
```
Comment (elkavitaly) [inline OneSchedule/SetWebHookService.cs]: Use ternary operator

File: OneSchedule.Settings/WebHookSettings.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿namespace OneSchedule.Settings
+{
+    public class WebHookSettings
+    {
+        public string Uri { get; set; }
+
+        public bool IsEnable { get; set; }
```
Comment (elkavitaly) [inline OneSchedule.Settings/WebHookSettings.cs]: rename to **IsEnabled**

File: OneSchedule/appsettings.json [hunk 11]
```diff
@@ -35,5 +35,10 @@
 
   "Quartz": {
     "quartz.scheduler.instanceName": "Quartz ASP.NET Core Sample Scheduler"
-  }
+  },
+
+  "WebHook": {
+    "IsUsed": true,
```
Comment (elkavitaly) [inline OneSchedule/appsettings.json]: Update config values according to property name

**Review by elkavitaly — CHANGES_REQUESTED** [review]

File: OneSchedule/SetWebHookService.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+namespace OneSchedule
+{
+    public class SetWebHookService : IHostedService
+    {
+        private readonly IOptions<WebHookSettings> _webHookOptions;
+        private readonly ITelegramBotClient _bot;
+
+        public SetWebHookService(IOptions<WebHookSettings> webHookOptions, ITelegramBotClient bot)
+        {
+            _webHookOptions = webHookOptions;
```
Comment (elkavitaly) [inline OneSchedule/SetWebHookService.cs:17]: It is usually better to have a field of settings class itself instead of IOptions<T>

**Review by elkavitaly — APPROVED** [review]

**Review by gitrepository100 — APPROVED** [review]
Approved

**Pull request merged**